File: backend-python/app/routers/providers.py
"""
Provider API Routes
Implements Phase 0 endpoints: GET /providers, POST /providers/test, GET /models
"""
from typing import Optional
from fastapi import APIRouter, HTTPException
from pydantic import BaseModel

from ..services.provider_service import provider_service


router = APIRouter(prefix="/providers", tags=["providers"])


class TestProviderRequest(BaseModel):
    """Request body for testing a provider"""
    id: str
    api_key: Optional[str] = None
    base_url: Optional[str] = None


class TestProviderResponse(BaseModel):
    """Response for provider test"""
    success: bool
    models: Optional[list] = None


class ModelsResponse(BaseModel):
    """Response for models endpoint"""
    models: list
    provider: str


@router.get("")
async def list_providers():
    """
    GET /api/providers
    List all available LLM providers
    """
    providers = await provider_service.list_providers()
    return {"providers": [p.model_dump() for p in providers]}


@router.post("/test")
async def test_provider(request: TestProviderRequest):
    """
    POST /api/providers/test
    Test connection to a provider
    """
    # If custom baseUrl provided, update temporarily
    if request.base_url:
        provider_service.update_provider(request.id, base_url=request.base_url)
    
    result = await provider_service.test_provider(request.id, request.api_key)
    
    if result.success:
        return {"success": True, "models": result.models}
    else:
        raise HTTPException(
            status_code=502,
            detail={
                "code": "LLM_ERROR",
                "message": result.error or "Provider connection failed",
            }
        )


@router.get("/models")
async def get_models(provider: str = "llama.cpp"):
    """
    GET /api/models
    Get available models for a provider
    """
    models = await provider_service.get_models(provider)
    return {"models": models, "provider": provider}

File: backend-nodejs/src/routes/providers.ts
import { Router, Request, Response } from 'express';
import { providerService } from '../services/providerService';

const router = Router();

/**
 * GET /api/providers
 * List all available LLM providers
 */
router.get('/', async (_req: Request, res: Response) => {
    try {
        const providers = await providerService.listProviders();
        res.json({ providers });
    } catch (error) {
        console.error('Error listing providers:', error);
        res.status(500).json({
            error: {
                code: 'SERVER_ERROR',
                message: 'Failed to list providers',
            },
        });
    }
});

/**
 * POST /api/providers/test
 * Test connection to a provider
 * Body: { id: string, apiKey?: string, baseUrl?: string }
 */
router.post('/test', async (req: Request, res: Response) => {
    const { id, apiKey, baseUrl } = req.body;

    if (!id) {
        return res.status(400).json({
            error: {
                code: 'VALIDATION_ERROR',
                message: 'Provider id is required',
                details: { field: 'id' },
            },
        });
    }

    try {
        // If custom baseUrl provided, update temporarily
        if (baseUrl) {
            providerService.updateProvider(id, { baseUrl });
        }

        const result = await providerService.testProvider(id, apiKey);

        if (result.success) {
            res.json({
                success: true,
                models: result.models,
            });
        } else {
            res.status(502).json({
                error: {
                    code: 'LLM_ERROR',
                    message: result.error || 'Provider connection failed',
                },
            });
        }
    } catch (error) {
        console.error('Error testing provider:', error);
        res.status(500).json({
            error: {
                code: 'SERVER_ERROR',
                message: 'Failed to test provider',
            },
        });
    }
});

/**
 * GET /api/models
 * Get available models for the active/specified provider
 * Query: ?provider=openai (optional)
 */
router.get('/models', async (req: Request, res: Response) => {
    const providerId = (req.query.provider as string) || 'llama.cpp';

    try {
        const models = await providerService.getModels(providerId);
        res.json({ models, provider: providerId });
    } catch (error) {
        console.error('Error fetching models:', error);
        res.status(502).json({
            error: {
                code: 'LLM_ERROR',
                message: 'Failed to fetch models from provider',
            },
        });
    }
});

export default router;
